### ISSUE TITLE
OpenRefine 3.8.2 Mac Application Not Notarized Correctly

### ISSUE THREAD

**Issue by uurazzle** [OpenRefine/OpenRefine#6713]
Hi:

Thanks for developing and making OpenRefine available to the community.

There is an OpenRefine 3.8.2 Mac application that is not notarized correctly. The app or component was signed with an Apple-issued Developer ID certificate but does not appear to be notarized. Since the exceptions described above for the Developer ID status are not met, macOS will not allow this code to run (unless it is found to be notarized when Gatekeeper runs).

To verify if a Mac application is notarized, you can use the spctl command in Terminal. Here is the command:

```
spctl -a -v /path/to/application.app
```

For example, executing the command with OpenRefine 3.8.2 outputs...

```
spctl -a -v /Volumes/OpenRefine/OpenRefine.app   
/Volumes/OpenRefine/OpenRefine.app: rejected
source=Unnotarized Developer ID
```

Additional context

Here is some interesting general information to checkout and share with your macOS developer on resolving Gatekeeper problems on macOS.

Resolving Gatekeeper Problems | Apple Developer Forums:

The post titled "Resolving Gatekeeper Problems" on the Apple Developer Forums, written by Quinn "The Eskimo!" from Developer Technical Support at Apple is a comprehensive guide addressing common issues related to Gatekeeper on macOS. Gatekeeper is a security feature designed to ensure that only trusted software runs on a user's Mac, and the post focuses on helping developers troubleshoot and resolve issues that may arise in this context.

The post identifies four common Gatekeeper problems that developers may encounter:

App blocked by a dangling load command path.
Broken code signature.
Lack of notarization.
Command-line tool blocked by Gatekeeper.
For each of these issues, the post provides detailed steps and guidance on how developers can resolve them. The emphasis is on the importance of passing Gatekeeper checks to maintain customer trust and avoid potential loss of customers.
Key points covered in the post include:

• Verification of Code Signature: Developers are advised to use the codesign tool to verify that their code is signed correctly. The post provides examples of command-line usage to check for issues such as missing or invalid sealed resources.
• Notarization Issues: Gatekeeper requires that apps be notarized, and the post guides developers on how to identify and resolve notarization problems. It includes information on checking system logs for specific entries related to notarization issues.
• Hash Mismatch: In cases where there's a hash mismatch, the post provides guidance based on the file type (e.g., zip archive, signed disk image, installer package) and recommends specific actions to address the problem.
• Command-line Tool Blocking Bug: A known bug in macOS is acknowledged, where double-clicking a command-line tool in Finder may lead to it being blocked by Gatekeeper. Workarounds, such as embedding the tool in an application or using an installer package, are suggested.

Throughout the post, there are references to Apple's documentation and resources related to code signing and notarization, providing developers with additional information for a deeper understanding.

https://forums.developer.apple.com/forums/thread/706379

**Comment by wetneb** [comment]
The notarization indeed failed:
```
[INFO]      [exec] Conducting pre-submission checks for openrefine-mac-3.8.2.dmg and initiating connection to the Apple notary service...
[INFO]      [exec] Error: HTTP status code: 403. A required agreement is missing or has expired. This request requires an in-effect agreement that has not been signed or has expired. Ensure your team has signed the necessary legal agreements and that they are not expired.
```
I have notified CS&S who take care of our Apple developer account. They need to accept new terms & conditions before we can continue to sign packages.

**Comment by wetneb** [comment]
Update on this: CS&S has lost access to the Apple developer account after staff changes, so it will take more time for them to recover the access and renew the agreement. They are looking into it.

**Comment by magdmartin** [comment]
If CS&S is unable to recover their account, I will create a new one as a project manager. I will set up a dedicated email address and forward it to the personal email of the individual who needs access to the account (Apple is using the email to enforce MFA). 

@tfmorris, do you have a timeline for releasing version 3.9? I want to ensure that we have a new certificate by then.

**Comment by wetneb** [comment]
> Apple is using the email to enforce MFA

A shared email address is not enough, since to log in you also get an SMS with a code.
But in practice developers don't need to access this regularly (not even for releases because the certificates are stored in the CI). All that needs doing (by the account holder) is to regularly accept the new terms & conditions on the web portal.

**Comment by tfmorris** [comment]
> do you have a timeline for releasing version 3.9? I want to ensure that we have a new certificate by then.

I'm targeting the first beta for the second week of October, but I'd be happy to have things in place earlier.

**Comment by magdmartin** [comment]
Here’s an update on the situation with CS&S. We met with their team on September 27 to discuss our options. At that time, we decided to wait until mid-October for the new Fiscal Sponsorship Program Director to take over the account. 

Now, as of October 17, I have emailed the new director to request an update. 

If the development team prefers not to wait any longer, I can begin the registration process with Apple on Monday.

**Comment by tfmorris** [comment]
Thanks for the update, Martin. From my point of view, $99/yr is a
pittance compared to the tens of thousands of dollars spent in Berlin,
so I'd much rather be in charge of our own destiny, but the bigger
question, at least for me, is the logistics of maintaining the account
on behalf of the project and delegating access to those who need to
sign releases. I don't know what that entails.

From my point of view, the priorities are (in priority order):
- a signature that meets Apple's rules (as broken as they are)
- a signature which is verifiably identifiable with the project
- a signature which is stable over time from release to release

I've started the prep work for the 3.9 release and, while I don't need
the credentials for the beta, it would be a good dry run.

Tom

**Comment by magdmartin** [comment]
I agree that $99 per year is a minimal expense compared to others. My main concern is that the account will remain tied to an individual rather than an institution. I believe that CS&S serves as a stable home base; for example, they already manage the domain openrefine.org. CS&S experienced a challenging six months due to significant staff turnover, which has led us to this situation (both the primary Apple account holder and its backup at CS&S left). However, from my conversations with them, I trust that they are reorganizing to prevent such issues in the future.

Regarding the signature @wetneb summarized it well in [his previous comment](https://github.com/OpenRefine/OpenRefine/issues/6713#issuecomment-2364574560)
> But in practice developers don't need to access this regularly (not even for releases because the certificates are stored in the CI). All that needs doing (by the account holder) is to regularly accept the new terms & conditions on the web portal.

Here is the latest update from CS&S (1h ago)
> A new ticket has already been opened with Apple for this matter and all the documentation is ready for submission.  Just awaiting the response from Apple which should soon (tomorrow i hope).

I will continue to update as I know more.

**Comment by tfmorris** [comment]
On Thu, Oct 17, 2024 at 11:34 PM Martin Magdinier ***@***.***>
wrote:

> A new ticket has already been opened with Apple for this matter and all
> the documentation is ready for submission. Just awaiting the response from
> Apple which should soon (tomorrow i hope).
>
> That sounds like good progress, so happy to wait.

Tom

**Comment by magdmartin** [comment]
@wetneb  Do you need the Apple certification when releasing [3.8.3 release?](https://forum.openrefine.org/t/releasing-3-8-3/1814)? In this case, we should start getting our own account and not wait for CS&S.

**Comment by wetneb** [comment]
Well, if it's activated by then, the release will be notarized, if not, it will not be notarized… I wouldn't wait longer to deliver those (actual) security fixes. We could do a 3.8.4 later if we recover notarization I guess.

**Comment by wetneb** [comment]
We are now able to notarize our builds again.

**Comment by kryptonit1** [comment]
When is the next release scheduled to be released? It would be nice to get access to a release that is signed and notarized as soon as possible.

**Comment by wetneb** [comment]
we can ask @tfmorris if he wants to do a 3.8.6 for it.

**Comment by tfmorris** [comment]
We have some other fixes going out in a 3.8.6, so it will include the
notarization when it's released.

**Comment by wetneb** [comment]
This has been fixed in 3.8.7.
